### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by alexsporn**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+COPY ./config.json /app/config.json
+
+############################
+# Image
+############################
+# https://console.cloud.google.com/gcr/images/distroless/global/cc-debian11
+# using distroless cc "nonroot" image, which includes everything in the base image (glibc, libssl and openssl)
+FROM gcr.io/distroless/cc-debian11:nonroot
+
+EXPOSE 9091/tcp
```
Comment (alexsporn) [inline Dockerfile]: This can be removed

File: pkg/coordinator/coordinator.go [hunk 2]
```diff
@@ -0,0 +1,591 @@
...
+// BackPressureFunc is a function which tells the Coordinator
+// to stop issuing milestones and checkpoints under high load.
+type BackPressureFunc func() bool
+
+// SendMessageFunc is a function which sends a message to the network.
+type SendMessageFunc = func(message *iotago.Message, msIndex ...milestone.Index) error
+
+type LatestMilestone struct {
+	Index       milestone.Index
+	Timestamp   uint64
```
Comment (muXxer) [inline pkg/coordinator/coordinator.go]: Why don't we use uint32 here?

File: pkg/coordinator/coordinator.go [hunk 3]
```diff
@@ -0,0 +1,591 @@
...
+			return ErrNetworkBootstrapped
+		}
+
+		if startIndex == 0 {
+			// start with milestone 1 at least
+			startIndex = 1
+		}
+
+		if latestMilestone.Index != startIndex-1 {
+			return fmt.Errorf("previous milestone does not match latest milestone in node! previous: %d, database: %d", startIndex-1, latestMilestone.Index)
```
Comment (muXxer) [inline pkg/coordinator/coordinator.go]: Replace second "database" as well.

File: pkg/coordinator/coordinator.go [hunk 4]
```diff
@@ -0,0 +1,591 @@
...
+		state := &State{}
+		state.LatestMilestoneMessageID = latestMilestoneMessageID
+		state.LatestMilestoneID = latestMilestoneID
+		state.LatestMilestoneIndex = startIndex - 1
+		state.LatestMilestoneTime = time.Now()
+
+		coo.state = state
+		coo.bootstrapped = false
+
+		coo.LogInfof("bootstrapping Coordinator at %d", startIndex)
```
Comment (muXxer) [inline pkg/coordinator/coordinator.go]: We always used lower-case for "Coordinator".

File: pkg/coordinator/coordinator.go [hunk 5]
```diff
@@ -0,0 +1,591 @@
...
+	coo.state = &State{}
+	if err := utils.ReadJSONFromFile(coo.opts.stateFilePath, coo.state); err != nil {
+		return err
+	}
+
+	if latestMilestone.Index != coo.state.LatestMilestoneIndex {
+		return fmt.Errorf("previous milestone does not match latest milestone in node. previous: %d, INX: %d", coo.state.LatestMilestoneIndex, latestMilestone.Index)
+	}
+
+	coo.LogInfof("resuming Coordinator at %d", latestMilestone.Index)
```
Comment (muXxer) [inline pkg/coordinator/coordinator.go]: We always used lower-case for "Coordinator".

File: pkg/coordinator/coordinator.go [hunk 6]
```diff
@@ -0,0 +1,591 @@
...
+}
+
+type IsNodeSyncedFunc = func() bool
+
+type MilestoneMerkleProof struct {
+	PastConeMerkleProof  *MerkleTreeHash
+	InclusionMerkleProof *MerkleTreeHash
+}
+
+type ComputeMerkleTreeHashFunc = func(ctx context.Context, index milestone.Index, timestamp uint64, parents hornet.MessageIDs, lastMilestoneID iotago.MilestoneID) (*MilestoneMerkleProof, error)
```
Comment (muXxer) [inline pkg/coordinator/coordinator.go]: timestamp uint32?

File: pkg/coordinator/coordinator.go [hunk 7]
```diff
@@ -0,0 +1,591 @@
...
+	// get receipt data in case migrator is enabled
+	var receipt *iotago.ReceiptMilestoneOpt
+	if coo.migratorService != nil {
+		receipt = coo.migratorService.Receipt()
+		if receipt != nil {
+			if err := coo.migratorService.PersistState(true); err != nil {
+				return common.CriticalError(fmt.Errorf("unable to persist migrator state before send: %w", err))
+			}
+
+			if coo.treasuryOutputFunc == nil {
```
Comment (muXxer) [inline pkg/coordinator/coordinator.go]: Can we move this check to the init of the coo?

File: pkg/coordinator/coordinator.go [hunk 8]
```diff
@@ -0,0 +1,591 @@
...
+	coo.state.LatestMilestoneMessageID = hornet.MessageIDFromArray(*latestMilestoneMessageID)
+	coo.state.LatestMilestoneID = *milestoneID
+	coo.state.LatestMilestoneIndex = newMilestoneIndex
+	coo.state.LatestMilestoneTime = newMilestoneTimestamp
+
+	if err := utils.WriteJSONToFile(coo.opts.stateFilePath, coo.state, 0660); err != nil {
+		return common.CriticalError(fmt.Errorf("failed to update coordinator state file: %w", err))
+	}
+
+	coo.Events.IssuedMilestone.Trigger(coo.state.LatestMilestoneIndex, coo.state.LatestMilestoneMessageID)
```
Comment (muXxer) [inline pkg/coordinator/coordinator.go]: Maybe we should add the `milestoneID` as well to the event?

File: main.go [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+	node.Run(
+		node.WithInitPlugin(app.InitPlugin),
+		node.WithCorePlugins([]*node.CorePlugin{
+			inx.CorePlugin,
+			coordinator.CorePlugin,
+			gracefulshutdown.CorePlugin,
+		}...),
+		node.WithPlugins([]*node.Plugin{
+			migrator.Plugin,
+			//prometheus.Plugin,
```
Comment (muXxer) [inline main.go:22]: Please enable :)

File: go.mod [hunk 10]
```diff
@@ -1,3 +1,124 @@
 module github.com/gohornet/inx-coordinator
 
 go 1.18
+
+require (
+	github.com/bits-and-blooms/bitset v1.2.2
+	github.com/gohornet/hornet v1.1.0-rc1.0.20220412082038-f46fe2f1eaa4
```
Comment (muXxer) [inline go.mod]: That should point to the stardust code.

File: tools/coordinator_fix_state.go [hunk 11]
```diff
@@ -0,0 +1,127 @@
+package tools
+
+/*
+import (
```
Comment (muXxer) [inline tools/coordinator_fix_state.go:4]: I think this tool should be part of the repo, but maybe we create another binary for tooling.

Should we do the same for hornet tools maybe.

Btw, this tool only works with hornet databases :carlos:

File: config.json [hunk 12]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (muXxer) [inline config.json:1]: Should the config file really be empty?

File: prometheus/migrator.go [hunk 13]
```diff
@@ -0,0 +1,61 @@
+package prometheus
+
+/*
+import (
```
Comment (muXxer) [inline prometheus/migrator.go]: Please add that code :)

File: prometheus/coordinator.go [hunk 14]
```diff
@@ -0,0 +1,107 @@
+package prometheus
+
+/*
+import (
+	"github.com/prometheus/client_golang/prometheus"
```
Comment (muXxer) [inline prometheus/coordinator.go]: Please add that code :)

File: plugin/migrator/plugin.go [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+
+	"github.com/gohornet/hornet/pkg/common"
+	validator "github.com/gohornet/hornet/pkg/model/migrator"
+	"github.com/gohornet/hornet/pkg/model/utxo"
+	"github.com/gohornet/hornet/pkg/node"
+	"github.com/gohornet/hornet/pkg/shutdown"
+	"github.com/gohornet/inx-coordinator/pkg/migrator"
+	"github.com/iotaledger/hive.go/configuration"
+	"github.com/iotaledger/hive.go/timeutil"
+	"github.com/iotaledger/iota.go/api"
```
Comment (muXxer) [inline plugin/migrator/plugin.go]: Should we name the package `legacyapi`?

File: plugin/inx/plugin.go [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+	"github.com/gohornet/hornet/pkg/node"
+	"github.com/gohornet/hornet/pkg/shutdown"
+	"github.com/gohornet/inx-coordinator/pkg/nodebridge"
+	"github.com/gohornet/inx-coordinator/plugin/migrator"
+	"github.com/iotaledger/hive.go/configuration"
+	inx "github.com/iotaledger/inx/go"
+)
+
+func init() {
+	CorePlugin = &node.CorePlugin{
```
Comment (muXxer) [inline plugin/inx/plugin.go]: Why don't we have the same folder structure like in hornet?
"core" for core "plugins" and "plugins" for optional ones?

File: plugin/inx/plugin.go [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+		CorePlugin.LogPanic(err)
+	}
+}
+
+func run() {
+	if err := CorePlugin.Daemon().BackgroundWorker("INX", func(ctx context.Context) {
+		deps.NodeBridge.Start(ctx)
+		<-ctx.Done()
+		deps.Connection.Close()
+	}, shutdown.PriorityCloseDatabase); err != nil {
```
Comment (muXxer) [inline plugin/inx/plugin.go]: This should get its own priority group.

File: plugin/coordinator/plugin.go [hunk 18]
```diff
@@ -0,0 +1,552 @@
...
+		return nil
+	}
+	return deps.Coordinator.Events
+}
+
+func configureEvents() {
+	// pass all new solid messages to the selector
+	onMessageSolid = events.NewClosure(func(metadata *inx.MessageMetadata) {
+
+		if metadata.GetShouldReattach() {
```
Comment (muXxer) [inline plugin/coordinator/plugin.go]: Can we add `GetBelowMaxDepth` to the inx interface?
Even if it uses exactly the same underlying flag maybe?
I think that makes the code easier to read.

File: plugin/app/app.go [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+	"go.uber.org/dig"
+
+	"github.com/gohornet/hornet/pkg/node"
+	"github.com/iotaledger/hive.go/configuration"
+	"github.com/iotaledger/hive.go/logger"
+)
+
+var (
+	// configs
+	nodeConfig = configuration.New()
```
Comment (muXxer) [inline plugin/app/app.go]: inx-coordinator doesn't seem to have any version number?

File: plugin/app/app.go [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+
+	InitPlugin *node.InitPlugin
+)
+
+func init() {
+	InitPlugin = &node.InitPlugin{
+		Pluggable: node.Pluggable{
+			Name:           "App",
+			Params:         params,
+			InitConfigPars: initConfigPars,
```
Comment (muXxer) [inline plugin/app/app.go]: Why did you remove `provide`and `configure`?
I think that was a nice way to show the name of the app and the version, and that plugins are initialized.

File: plugin/app/app.go [hunk 21]
```diff
@@ -0,0 +1,166 @@
...
+
+	var flagSetsToParse = configFlagSets
+	flagSetsToParse["config_files"] = configFilesFlagSet
+
+	parseFlags(flagSetsToParse)
+
+	if err = loadCfg(configFlagSets); err != nil {
+		return nil, err
+	}
+	if err := logger.InitGlobalLogger(nodeConfig); err != nil {
```
Comment (muXxer) [inline plugin/app/app.go]: Can you add the following code before loading the logger?

```go
	if err = nodeConfig.SetDefault(logger.ConfigurationKeyDisableCaller, true); err != nil {
		panic(err)
	}
```
Otherwise the stack traces are huge.

File: plugin/app/app.go [hunk 22]
```diff
@@ -0,0 +1,166 @@
...
+		panic(err)
+	}
+
+	printConfig(maskedKeys)
```
Comment (muXxer) [inline plugin/app/app.go]: Maybe print the app name and version here again.

File: pkg/nodebridge/node_bridge.go [hunk 23]
```diff
@@ -0,0 +1,313 @@
...
+func (n *NodeBridge) KeyManager() *keymanager.KeyManager {
+	keyManager := keymanager.New()
+	for _, keyRange := range n.ProtocolParameters.GetMilestoneKeyRanges() {
+		keyManager.AddKeyRange(keyRange.GetPublicKey(), milestone.Index(keyRange.GetStartIndex()), milestone.Index(keyRange.GetEndIndex()))
+	}
+	return keyManager
+}
+
+func (n *NodeBridge) Start(ctx context.Context) {
+	go n.listenToConfirmedMilestone(ctx)
```
Comment (muXxer) [inline pkg/nodebridge/node_bridge.go]: How do we detect that the underlying stream failed?
Show we do proper error handling here with channels etc?
Or is this no problem at all?

File: pkg/nodebridge/node_bridge.go [hunk 24]
```diff
@@ -0,0 +1,313 @@
...
+	}
+
+	return n.latestMilestone.GetMilestoneIndex() == n.confirmedMilestone.GetMilestoneIndex()
+}
+
+func (n *NodeBridge) LatestMilestone() *coordinator.LatestMilestone {
+	n.isSyncedMutex.RLock()
+	defer n.isSyncedMutex.RUnlock()
+	return &coordinator.LatestMilestone{
+		Index:     milestone.Index(n.latestMilestone.GetMilestoneIndex()),
```
Comment (muXxer) [inline pkg/nodebridge/node_bridge.go:135]: `uint32`

File: pkg/nodebridge/node_bridge.go [hunk 25]
```diff
@@ -0,0 +1,313 @@
...
+		return nil, errors.New("haven't received any treasury outputs yet")
+	}
+
+	return &coordinator.LatestTreasuryOutput{
+		MilestoneID: n.latestTreasuryOutput.UnwrapMilestoneID(),
+		Amount:      n.latestTreasuryOutput.GetAmount(),
+	}, nil
+}
+
+func (n *NodeBridge) ComputeMerkleTreeHash(ctx context.Context, msIndex milestone.Index, msTimestamp uint32, parents hornet.MessageIDs, lastMilestoneID iotago.MilestoneID) (*coordinator.MilestoneMerkleProof, error) {
```
Comment (muXxer) [inline pkg/nodebridge/node_bridge.go]: `timestamp` => `uint32`

File: pkg/coordinator/coordinator.go [hunk 26]
```diff
@@ -0,0 +1,587 @@
...
+	IssuedMilestone *events.Event
+	// SoftError is triggered when a soft error is encountered.
+	SoftError *events.Event
+	// QuorumFinished is triggered after a coordinator quorum call was finished.
+	QuorumFinished *events.Event
+}
+
+type IsNodeSyncedFunc = func() bool
+
+type MilestoneMerkleProof struct {
```
Comment (muXxer) [inline pkg/coordinator/coordinator.go]: `Proof` or `Roots` or `Hashes`:carlos:

File: pkg/coordinator/debug_node_api.go [hunk 27]
```diff
@@ -0,0 +1,73 @@
...
+		return nil, fmt.Errorf("unknown merkle tree hash length (%d)", len(confirmedMerkleRootBytes))
+	}
+
+	appliedMerkleRootBytes, err := iotago.DecodeHex(res.AppliedMerkleRoot)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(appliedMerkleRootBytes) != iotago.MilestoneMerkleProofLength {
+		return nil, fmt.Errorf("unknown merkle tree hash length (%d)", len(appliedMerkleRootBytes))
```
Comment (muXxer) [inline pkg/coordinator/debug_node_api.go]: maybe add "applied" to the error message

File: pkg/coordinator/debug_node_api.go [hunk 28]
```diff
@@ -0,0 +1,73 @@
...
+		return nil, err
+	}
+
+	confirmedMerkleRootBytes, err := iotago.DecodeHex(res.ConfirmedMerkleRoot)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(confirmedMerkleRootBytes) != iotago.MilestoneMerkleProofLength {
+		return nil, fmt.Errorf("unknown merkle tree hash length (%d)", len(confirmedMerkleRootBytes))
```
Comment (muXxer) [inline pkg/coordinator/debug_node_api.go]: maybe add "confirmed" to the error message

**Review by muXxer — CHANGES_REQUESTED** [review]
Looks good so far!

The coo code had high coverage of comments to every function etc before. Maybe you could add this as well :trollface:

File: pkg/nodebridge/node_bridge.go (hunk 25 above)
Comment (alexsporn) [reply pkg/nodebridge/node_bridge.go]: huh? I see `uint32` here already

File: config.json (hunk 12 above)
Comment (alexsporn) [reply config.json:1]: Yes, because the defaults are in the code. It is kind of stupid to repeat the defaults imho.

File: plugin/coordinator/plugin.go (hunk 18 above)
Comment (alexsporn) [reply plugin/coordinator/plugin.go]: This MessageMetadata contains the same thing we expose over the API and MQTT. Do we want to add special fields to INX only?

File: pkg/nodebridge/node_bridge.go (hunk 23 above)
Comment (alexsporn) [reply pkg/nodebridge/node_bridge.go]: the context gets cancelled

File: plugin/coordinator/plugin.go (hunk 18 above)
Comment (muXxer) [reply plugin/coordinator/plugin.go]: For me INX is a level deeper than API or MQTT.
So yeah, better add a special flag for this and define on node side what "belowMaxDepth" really means.

File: config.json (hunk 12 above)
Comment (muXxer) [reply config.json:1]: Should we add a `config_template.json` to our repos then?
I hate to look into the code what options are possible. And explaining this to other people is even worse :D

File: core/inx/params.go [hunk 29]
```diff
@@ -0,0 +1,23 @@
...
+
+import (
+	flag "github.com/spf13/pflag"
+
+	"github.com/gohornet/hornet/pkg/node"
+)
+
+const (
+	// CfgINXAddress the INX address to which to connect to.
+	CfgINXAddress = "inx.address"
```
Comment (muXxer) [inline core/inx/params.go:11]: Should be `"inx.bindAddress"`

File: main.go (hunk 9 above)
Comment (alexsporn) [reply main.go:22]: Lets do it with #3

File: go.mod (hunk 10 above)
Comment (alexsporn) [reply go.mod]: 1.1.0-rc1 is what go makes itself when I point to to the commit in the branch 🤷

File: prometheus/migrator.go (hunk 13 above)
Comment (alexsporn) [reply prometheus/migrator.go]: Lets do it with #3

File: prometheus/coordinator.go (hunk 14 above)
Comment (alexsporn) [reply prometheus/coordinator.go]: Lets do it with #3

File: tools/coordinator_fix_state.go (hunk 11 above)
Comment (alexsporn) [reply tools/coordinator_fix_state.go:4]: Let's decide that in #4

File: core/app/app.go [hunk 30]
```diff
@@ -14,12 +14,15 @@ import (
...
+	Version = "0.1.0"
+
 	// configs
-	nodeConfig = configuration.New()
+	appConfig = configuration.New()
 
 	// config file flags
 	configFilesFlagSet = flag.NewFlagSet("config_files", flag.ContinueOnError)
-	nodeCfgFilePath    = configFilesFlagSet.StringP(CfgConfigFilePathNodeConfig, "c", "config.json", "file path of the config file")
+	nodeCfgFilePath    = configFilesFlagSet.StringP(CfgConfigFilePathAppConfig, "c", "config.json", "file path of the config file")
```
Comment (muXxer) [inline core/app/app.go]:
```suggestion
	appCfgFilePath    = configFilesFlagSet.StringP(CfgConfigFilePathAppConfig, "c", "config.json", "file path of the config file")
```

File: core/app/app.go [hunk 31]
```diff
@@ -53,40 +57,46 @@ func initialize(params map[string][]*flag.FlagSet, maskedKeys []string) (*node.I
...
+		DisabledPlugins: appConfig.Strings(CfgAppDisablePlugins),
 	}, nil
 }
 
 // parses the configuration and initializes the global logger.
 func loadCfg(flagSets map[string]*flag.FlagSet) error {
 
-	if hasFlag(flag.CommandLine, CfgConfigFilePathNodeConfig) {
+	if hasFlag(flag.CommandLine, CfgConfigFilePathAppConfig) {
 		// node config file is only loaded if the flag was specified
```
Comment (muXxer) [inline core/app/app.go]:
```suggestion
		// app config file is only loaded if the flag was specified
```

File: core/inx/params.go (hunk 29 above)
Comment (alexsporn) [reply core/inx/params.go:11]: no, we normally use bindAddress when we are exposing stuff

Comment (muXxer) [reply core/inx/params.go:11]: lol, right.
My fault.
But then the template config file is also wrong :runner:

File: plugin/coordinator/plugin.go (hunk 18 above)
Comment (muXxer) [reply plugin/coordinator/plugin.go]: Will be done with #6

File: pkg/nodebridge/node_bridge.go (hunk 23 above)
Comment (muXxer) [reply pkg/nodebridge/node_bridge.go]: Lets do that in a clean way with #7

**Review by muXxer — APPROVED** [review]
